lldb/source/Plugins/InstrumentationRuntime/BoundsSafety/InstrumentationRuntimeBoundsSafety.cpp

```diff
@@ -0,0 +1,336 @@
+    auto category = MaybeTrapReason.value().first;
+    auto message = MaybeTrapReason.value().second;
+
+    // TODO: Clang should probably be changed to emit the "Soft " prefix itself
+    std::string stop_reason;
+    llvm::raw_string_ostream ss(stop_reason);
+    ss << SOFT_TRAP_CATEGORY_PREFIX;
+    if (category.empty())
+      ss << "<empty category>";
+    else
+      ss << category;
+    if (!message.empty()) {
+      ss << ": " << message;
+    }
+    // Use computed stop-reason and assume the parent of `parent_sf` is the
+    // the place in the user's code where the call to the soft trap runtime
+    // originated.
+    return std::make_pair(stop_reason, parent_sf->GetFrameIndex() + 1);
+  }
+
+  std::pair<std::optional<std::string>, std::optional<uint32_t>>
+  ComputeStopReasonAndSuggestedStackFrameWithoutDebugInfo(ThreadSP thread_sp) {
+    auto softtrap_sf = thread_sp->GetStackFrameAtIndex(0);
+    if (!softtrap_sf)
+      return {};
+    llvm::StringRef trap_reason_func_name = softtrap_sf->GetFunctionName();
+
+    if (trap_reason_func_name == BOUNDS_SAFETY_SOFT_TRAP_MINIMAL) {
+      // This function has no arguments so there's no additional information
+      // that would allow us to identify the trap reason.
+      //
+      // Use the fallback stop reason and the current frame.
+      // While we "could" set the suggested frame to our parent (where the
+      // bounds check failed), doing this leads to very misleading output in
+      // LLDB. E.g.:
+      //
+      // ```
+      //     0x100003b40 <+104>: bl  0x100003d64    ; __bounds_safety_soft_trap
+      // ->  0x100003b44 <+108>: b   0x100003b48    ; <+112>
+      // ```
+      //
+      // This makes it look we stopped after finishing the call to
+      // `__bounds_safety_soft_trap` but actually we are in the middle of the
+      // call. To avoid this confusion just use the current frame.
+      return {};
+    }
+
+    // BOUNDS_SAFETY_SOFT_TRAP_S has one argument which is a pointer to a string
+    // describing the trap or a nullptr.
+    if (trap_reason_func_name != BOUNDS_SAFETY_SOFT_TRAP_S)
+      return {};
+
+    auto rc = thread_sp->GetRegisterContext();
+    if (!rc)
+      return {};
+
+    // Don't try for architectures where examining the first register won't
+    // work.
+    auto process = thread_sp->GetProcess();
+    if (!process)
+      return {};
+    switch (process->GetTarget().GetArchitecture().GetCore()) {
+    case ArchSpec::eCore_x86_32_i386:
+    case ArchSpec::eCore_x86_32_i486:
+    case ArchSpec::eCore_x86_32_i486sx:
+    case ArchSpec::eCore_x86_32_i686:
+      // Technically some x86 calling conventions do use a register for
+      // passing the first argument but let's ignore that for now.
+      return {};
+    default: {
+    }
+    };
+
+    // Examine the register for the first argument
+    auto *arg0_info = rc->GetRegisterInfo(
+        lldb::RegisterKind::eRegisterKindGeneric, LLDB_REGNUM_GENERIC_ARG1);
+    if (!arg0_info)
```

Comment:
Along this process there are a bunch of places where it could fail.  It might help your sanity later on if you logged where the failure occurs.  Then when this is failing on a machine or for a process you can't get access to you have more of a chance of figuring out what went wrong.

Reply:
TIL, there isn't an instrumentation runtime log channel, which there really should be.  So maybe better to do this as a follow-on.

Reply:
I've cherry-picked the upstream change into this PR and added logging.